File: server/src/database/schema.service.ts
import { Injectable } from '@nestjs/common';
import { DatabaseService } from './database.service';
import { RowDataPacket } from 'mysql2/promise';

const PRESET_BUSINESS_TABLES = ['platform_info', 'daily_active_stats', 'user_behavior_log'];

interface SchemaCacheEntry {
	prompt: string;
	timestamp: number;
}

const CACHE_TTL_MS = 60 * 1000; // 1 分钟缓存

@Injectable()
export class SchemaService {
	private readonly cache = new Map<string, SchemaCacheEntry>();

	constructor(private readonly db: DatabaseService) { }

	async getDatabaseSchemaPrompt(datasourceId: number | null, userId: number): Promise<string> {
		const cacheKey = `${datasourceId || 'local'}_${userId}`;
		const cached = this.cache.get(cacheKey);
		if (cached && Date.now() - cached.timestamp < CACHE_TTL_MS) {
			return cached.prompt;
		}

		const isInternal = await this.isInternalDs(datasourceId);
		const uploadedTableNames = await this.getUserUploadedTableNames(datasourceId, userId);
		
		const allowedTables = isInternal 
			? [...PRESET_BUSINESS_TABLES, ...uploadedTableNames]
			: uploadedTableNames;

		if (allowedTables.length === 0) {
			return isInternal ? '当前数据库中没有可用的业务表。' : '当前数据源中还没有上传的表。';
		}

		const prompt = await this.fetchSchemaPromptFromLocal(allowedTables, uploadedTableNames);
		
		if (prompt) {
			this.cache.set(cacheKey, { prompt, timestamp: Date.now() });
		}
		return prompt;
	}

	clearUserCache(datasourceId: number | null, userId: number): void {
		const cacheKey = `${datasourceId || 'local'}_${userId}`;
		this.cache.delete(cacheKey);
	}

	async getAllowedTableNames(datasourceId: number | null, userId: number): Promise<string[]> {
		const isInternal = await this.isInternalDs(datasourceId);
		const uploadedTableNames = await this.getUserUploadedTableNames(datasourceId, userId);
		
		if (isInternal) {
			return [...PRESET_BUSINESS_TABLES, ...uploadedTableNames];
		}
		return uploadedTableNames;
	}

	private async isInternalDs(datasourceId: number | null): Promise<boolean> {
		if (!datasourceId) return true;
		const rows = await this.db.query<RowDataPacket[]>(
			'SELECT is_local FROM data_source WHERE id = ?',
			[datasourceId],
		);
		if (!rows.length) return true;
		return !!(rows[0] as any).is_local;
	}

	private async fetchSchemaPromptFromLocal(allowedTables: string[], uploadedTableNames: string[]): Promise<string> {
		const dbNameRows = await this.db.query<any[]>('SELECT DATABASE() AS db_name');
		const dbName = dbNameRows[0]?.db_name;
		if (!dbName) return '无法获取目标数据库名称。';

		const placeholders = allowedTables.map(() => '?').join(', ');
		const sql = `
			SELECT
				c.TABLE_NAME,
				t.TABLE_COMMENT,
				c.COLUMN_NAME,
				c.COLUMN_TYPE,
				c.COLUMN_COMMENT
			FROM information_schema.COLUMNS c
			JOIN information_schema.TABLES t
			  ON c.TABLE_NAME = t.TABLE_NAME AND c.TABLE_SCHEMA = t.TABLE_SCHEMA
			WHERE c.TABLE_SCHEMA = ?
			  AND c.TABLE_NAME IN (${placeholders})
			ORDER BY c.TABLE_NAME, c.ORDINAL_POSITION
		`;

		const rows = await this.db.query<any[]>(sql, [dbName, ...allowedTables]);
		return this.formatSchemaRows(rows, uploadedTableNames);
	}

	async getStructuredSchema(datasourceId: number | null, userId: number): Promise<any[]> {
		const isInternal = await this.isInternalDs(datasourceId);
		const uploadedTableNames = await this.getUserUploadedTableNames(datasourceId, userId);
		
		const allowedTables = isInternal 
			? [...PRESET_BUSINESS_TABLES, ...uploadedTableNames]
			: uploadedTableNames;

		if (allowedTables.length === 0) return [];

		const dbNameRows = await this.db.query<any[]>('SELECT DATABASE() AS db_name');
		const dbName = dbNameRows[0]?.db_name;
		if (!dbName) return [];

		const placeholders = allowedTables.map(() => '?').join(', ');
		const sql = `
			SELECT
				c.TABLE_NAME,
				t.TABLE_COMMENT,
				c.COLUMN_NAME,
				c.COLUMN_TYPE,
				c.COLUMN_COMMENT
			FROM information_schema.COLUMNS c
			JOIN information_schema.TABLES t
			  ON c.TABLE_NAME = t.TABLE_NAME AND c.TABLE_SCHEMA = t.TABLE_SCHEMA
			WHERE c.TABLE_SCHEMA = ?
			  AND c.TABLE_NAME IN (${placeholders})
			ORDER BY c.TABLE_NAME, c.ORDINAL_POSITION
		`;

		const rows = await this.db.query<any[]>(sql, [dbName, ...allowedTables]);
		const tablesMap = new Map<string, { tableName: string; display_name: string; isUploaded: boolean; fields: any[] }>();

		for (const row of rows) {
			const tableName = row.TABLE_NAME;
			if (!tablesMap.has(tableName)) {
				tablesMap.set(tableName, {
					tableName,
					display_name: row.TABLE_COMMENT || tableName,
					isUploaded: uploadedTableNames.includes(tableName),
					fields: [],
				});
			}
			const tableInfo = tablesMap.get(tableName)!;
			tableInfo.fields.push({
				name: row.COLUMN_NAME,
				type: row.COLUMN_TYPE,
				comment: row.COLUMN_COMMENT || '',
			});
		}

		return Array.from(tablesMap.values());
	}

	private async getUserUploadedTableNames(datasourceId: number | null, userId: number): Promise<string[]> {
		const actualDatasourceId = datasourceId ?? await this.getLocalDatasourceId();
		if (!actualDatasourceId) return [];

		const rows = await this.db.query<RowDataPacket[]>(
			'SELECT table_name FROM uploaded_table WHERE datasource_id = ? AND user_id = ?',
			[actualDatasourceId, userId],
		);
		return rows.map((r: any) => r.table_name);
	}

	private async getLocalDatasourceId(): Promise<number | null> {
		const rows = await this.db.query<RowDataPacket[]>(
			'SELECT id FROM data_source WHERE is_local = 1 LIMIT 1',
		);
		return rows.length > 0 ? (rows[0] as any).id : null;
	}

	private formatSchemaRows(rows: any[], uploadedTableNames: string[]): string {
		if (!rows || rows.length === 0) {
			return '当前数据库中没有可用的业务表。';
		}

		const tablesMap = new Map<string, { comment: string; columns: string[]; isUploaded: boolean }>();

		for (const row of rows) {
			const tableName = row.TABLE_NAME;
			if (!tablesMap.has(tableName)) {
				tablesMap.set(tableName, {
					comment: row.TABLE_COMMENT || '',
					columns: [],
					isUploaded: uploadedTableNames.includes(tableName),
				});
			}
			const tableInfo = tablesMap.get(tableName)!;
			const colDef = `  ${row.COLUMN_NAME} ${row.COLUMN_TYPE} -- ${row.COLUMN_COMMENT || ''}`;
			tableInfo.columns.push(colDef);
		}

		const promptLines: string[] = [];
		for (const [tableName, info] of tablesMap.entries()) {
			const label = info.isUploaded ? ' [用户上传]' : '';
			promptLines.push(`表名: ${tableName}${label}`);
			promptLines.push(`说明: ${info.comment}`);
			promptLines.push(`字段:\n${info.columns.join('\n')}`);
			promptLines.push('');
		}

		return promptLines.join('\n').trim();
	}
}

File: java-server/src/main/java/com/gistagent/database/SchemaService.java
package com.gistagent.database;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.gistagent.common.JdbcScalars;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;

/**
 * Manages DB schema visibility and prompt generation for the AI.
 *
 * <p>Port of the Nest SchemaService. Access control semantics: - internal/local datasource: preset
 * business tables + user's uploaded tables - external datasource: only user's uploaded tables
 */
@Service
public class SchemaService {

  private static final List<String> PRESET_BUSINESS_TABLES =
      List.of("platform_info", "daily_active_stats", "user_behavior_log");
  private static final Duration CACHE_TTL = Duration.ofMinutes(1);

  private final JdbcTemplate jdbc;
  private final Map<String, CacheEntry> cache = new ConcurrentHashMap<>();

  public SchemaService(JdbcTemplate jdbc) {
    this.jdbc = jdbc;
  }

  public String getDatabaseSchemaPrompt(Long datasourceId, long userId) {
    String cacheKey = cacheKey(datasourceId, userId);
    CacheEntry cached = cache.get(cacheKey);
    if (cached != null
        && Duration.between(cached.createdAt, Instant.now()).compareTo(CACHE_TTL) < 0) {
      return cached.prompt;
    }

    boolean internal = isInternalDs(datasourceId);
    List<String> uploaded = getUserUploadedTableNames(datasourceId, userId);
    List<String> allowed = internal ? merge(PRESET_BUSINESS_TABLES, uploaded) : uploaded;

    if (allowed.isEmpty()) {
      return internal ? "当前数据库中没有可用的业务表。" : "当前数据源中还没有上传的表。";
    }

    String prompt = fetchSchemaPromptFromLocal(allowed, uploaded);
    if (prompt != null && !prompt.isBlank()) {
      cache.put(cacheKey, new CacheEntry(prompt, Instant.now()));
    }
    return prompt;
  }

  public void clearUserCache(Long datasourceId, long userId) {
    cache.remove(cacheKey(datasourceId, userId));
  }

  public List<String> getAllowedTableNames(Long datasourceId, long userId) {
    boolean internal = isInternalDs(datasourceId);
    List<String> uploaded = getUserUploadedTableNames(datasourceId, userId);
    return internal ? merge(PRESET_BUSINESS_TABLES, uploaded) : uploaded;
  }

  public List<StructuredTable> getStructuredSchema(Long datasourceId, long userId) {
    boolean internal = isInternalDs(datasourceId);
    List<String> uploaded = getUserUploadedTableNames(datasourceId, userId);
    List<String> allowed = internal ? merge(PRESET_BUSINESS_TABLES, uploaded) : uploaded;
    if (allowed.isEmpty()) return List.of();

    String dbName = currentDatabaseName();
    if (dbName == null) return List.of();

    List<Map<String, Object>> rows = queryColumnRows(dbName, allowed);
    Map<String, StructuredTable> map = new LinkedHashMap<>();
    for (Map<String, Object> row : rows) {
      String tableName = (String) row.get("TABLE_NAME");
      StructuredTable table =
          map.computeIfAbsent(
              tableName,
              name ->
                  new StructuredTable(
                      name,
                      stringOrDefault(row.get("TABLE_COMMENT"), name),
                      uploaded.contains(name),
                      new ArrayList<>()));
      table
          .fields()
          .add(
              new StructuredField(
                  (String) row.get("COLUMN_NAME"),
                  (String) row.get("COLUMN_TYPE"),
                  stringOrDefault(row.get("COLUMN_COMMENT"), "")));
    }
    return new ArrayList<>(map.values());
  }

  private boolean isInternalDs(Long datasourceId) {
    if (datasourceId == null) return true;
    List<Map<String, Object>> rows =
        jdbc.queryForList("SELECT is_local FROM data_source WHERE id = ?", datasourceId);
    if (rows.isEmpty()) return true;
    return JdbcScalars.isTruthy(rows.get(0).get("is_local"));
  }

  private String fetchSchemaPromptFromLocal(List<String> allowed, List<String> uploaded) {
    String dbName = currentDatabaseName();
    if (dbName == null) return "无法获取目标数据库名称。";

    List<Map<String, Object>> rows = queryColumnRows(dbName, allowed);
    return formatSchemaRows(rows, uploaded);
  }

  private String currentDatabaseName() {
    List<String> names = jdbc.queryForList("SELECT DATABASE()", String.class);
    return names.isEmpty() ? null : names.get(0);
  }

  private List<Map<String, Object>> queryColumnRows(String dbName, List<String> allowedTables) {
    String placeholders = String.join(", ", Collections.nCopies(allowedTables.size(), "?"));
    String sql =
        """
				SELECT
					c.TABLE_NAME,
					t.TABLE_COMMENT,
					c.COLUMN_NAME,
					c.COLUMN_TYPE,
					c.COLUMN_COMMENT
				FROM information_schema.COLUMNS c
				JOIN information_schema.TABLES t
				  ON c.TABLE_NAME = t.TABLE_NAME AND c.TABLE_SCHEMA = t.TABLE_SCHEMA
				WHERE c.TABLE_SCHEMA = ?
				  AND c.TABLE_NAME IN (%s)
				ORDER BY c.TABLE_NAME, c.ORDINAL_POSITION
				"""
            .formatted(placeholders);

    Object[] params = new Object[allowedTables.size() + 1];
    params[0] = dbName;
    for (int i = 0; i < allowedTables.size(); i++) {
      params[i + 1] = allowedTables.get(i);
    }
    return jdbc.queryForList(sql, params);
  }

  private List<String> getUserUploadedTableNames(Long datasourceId, long userId) {
    Long actualId = datasourceId != null ? datasourceId : getLocalDatasourceId();
    if (actualId == null) return List.of();
    return jdbc.queryForList(
        "SELECT table_name FROM uploaded_table WHERE datasource_id = ?", String.class, actualId);
  }

  private Long getLocalDatasourceId() {
    List<Long> ids =
        jdbc.queryForList("SELECT id FROM data_source WHERE is_local = 1 LIMIT 1", Long.class);
    return ids.isEmpty() ? null : ids.get(0);
  }

  private String formatSchemaRows(List<Map<String, Object>> rows, List<String> uploadedTableNames) {
    if (rows == null || rows.isEmpty()) return "当前数据库中没有可用的业务表。";

    Map<String, TableInfo> map = new LinkedHashMap<>();
    for (Map<String, Object> row : rows) {
      String tableName = (String) row.get("TABLE_NAME");
      TableInfo info =
          map.computeIfAbsent(
              tableName,
              name ->
                  new TableInfo(
                      stringOrDefault(row.get("TABLE_COMMENT"), ""),
                      uploadedTableNames.contains(name),
                      new ArrayList<>()));
      info.columns()
          .add(
              "  %s %s -- %s"
                  .formatted(
                      row.get("COLUMN_NAME"),
                      row.get("COLUMN_TYPE"),
                      stringOrDefault(row.get("COLUMN_COMMENT"), "")));
    }

    return map.entrySet().stream()
        .map(
            e -> {
              String label = e.getValue().isUploaded() ? " [用户上传]" : "";
              return "表名: "
                  + e.getKey()
                  + label
                  + "\n"
                  + "说明: "
                  + e.getValue().comment()
                  + "\n"
                  + "字段:\n"
                  + String.join("\n", e.getValue().columns())
                  + "\n";
            })
        .collect(Collectors.joining("\n"))
        .trim();
  }

  private static List<String> merge(List<String> a, List<String> b) {
    List<String> out = new ArrayList<>(a.size() + b.size());
    out.addAll(a);
    out.addAll(b);
    return out;
  }

  private static String cacheKey(Long datasourceId, long userId) {
    return (datasourceId == null ? "local" : datasourceId.toString()) + "_" + userId;
  }

  private static String stringOrDefault(Object value, String fallback) {
    if (value == null) return fallback;
    String s = value.toString();
    return s.isEmpty() ? fallback : s;
  }

  private record CacheEntry(String prompt, Instant createdAt) {}

  private record TableInfo(String comment, boolean isUploaded, List<String> columns) {}

  public record StructuredTable(
      String tableName,
      @JsonProperty("display_name") String displayName,
      boolean isUploaded,
      List<StructuredField> fields) {}

  public record StructuredField(String name, String type, String comment) {}
}
